File: merge sort/merge_sort_class.cpp
#include <iostream>
using namespace std;

class MergeSortArray
{
	public:
		MergeSortArray();
		MergeSortArray(int size);
		~MergeSortArray();

		void Insert(int num);
		void Merge(int &start, int &middle, int &end);
		void Show();
		void Sort(int start, int &end);

	private:
		int capacity_;
		int end_;
		int size_;
		int start_;
		int *arr_;
};

MergeSortArray::MergeSortArray():
	MergeSortArray(0)
{
}

MergeSortArray::MergeSortArray(int size):
	capacity_(size*2),
	end_(size-1),
	size_(0),
	start_(0)
{
	arr_ = new int[size_];
}

MergeSortArray::~MergeSortArray()
{
    delete [] arr_;
}

void MergeSortArray::Insert(int num) {
	if (size_ < capacity_) {
		arr_[size_] = num;
        size_++;
	} else {
		capacity_ *= 2;
		int *array = new int[capacity_];
		for (int i = 0; i < size_; i++) {
			array[i] = arr_[i];
		}
		array[size_] = num;
		delete [] arr_;
		arr_ = array;
        size_++;
	}
}

void MergeSortArray::Merge(int &start, int &middle, int &end) {
	int front = middle - start + 1;
	int rear = end - middle;
	int left[front], right[rear];

	for (int i = start; i <= end; i++) {
		i <= middle ? left[i - start] = arr_[i] : right[i - middle - 1] = arr_[i];
	}

    for (int i = start, l = 0, r = 0; i <= end; i++) {
		if ((l < front) && (r < rear)) {
			arr_[i] = left[l] < right[r] ? left[l++] : right[r++];
		} else {
			arr_[i] = l < front ? left[l++] : right[r++];
		}
	}
}

void MergeSortArray::Show() {
	for (int i = 0; i < size_; i++) {
		cout << arr_[i] << (i == size_-1 ? "" : " ");
	}
	cout << endl;
}

void MergeSortArray::Sort(int start, int &end) {
	if (start < end) {
		int middle = (start + end) / 2;
		Sort(start, middle);
		Sort(middle+1, end);
		Merge(start, middle, end);
	}
}


int main()
{
	int n = 0;
	cin >> n;
	MergeSortArray numbers(n);
	for (int i = 0; i < n; i++) {
		int num = 0;
		cin >> num;
		numbers.Insert(num);
	}
	int start = 0;
	int end = n-1;
	numbers.Sort(start, end);
	numbers.Show();
}

File: merge sort/merge_sort_class.py
class MergeSortArray():
    """A class performing merge sort algorithm."""
    def __init__(self, size):
        """Initializing parameters when construct class."""
        self.arr = list()
        self.end = size - 1
        self.size = size
        self.start = 0

    def Insert(self, number):
        """Inserting number into array."""
        self.arr.append(number)

    def Merge(self, start, middle, end):
        """Merging two arrays by ascending."""
        front = middle - start + 1
        rear = end - middle
        left = [self.arr[index] for index in range(start, middle+1)]
        right = [self.arr[index] for index in range(middle+1, end+1)]

        l = 0
        r = 0
        for index in range(start, end+1):
            if l < front and r < rear:
                if left[l] < right[r]:
                    self.arr[index] = left[l]
                    l += 1
                else:
                    self.arr[index] = right[r]
                    r += 1
            else:
                if l < front:
                    self.arr[index] = left[l]
                    l += 1
                else:
                    self.arr[index] = right[r]
                    r += 1

    def Show(self):
        """Show every element in the array separate by space."""
        for index in range(len(self.arr)):
            print(str(self.arr[index]) if index == len(self.arr)-1 else str(self.arr[index]) + " ", end='')

    def Sort(self, start, end):
        """Sort array into two arrays."""
        if start < end:
            middle = int((start + end) / 2)
            self.Sort(start, middle)
            self.Sort(middle+1, end)
            self.Merge(start, middle, end)


n = int(input())
numbers = MergeSortArray(n)
for number in list(map(int, input().split())):
    numbers.Insert(number)

numbers.Sort(0, n-1)
numbers.Show()
        